docs: reflect the stateless protocol, server/discover, and session removal (#3069)
modelcontextprotocol/modelcontextprotocol: docs/docs/draft/learn/architecture.mdx

Reviewer: ```suggestion
for client in app.mcp_clients():
tools_response = await cient.list_tools()
available_tools.extend(tools_response.tools)
```

Reviewer: could maybe link to our guidance around using things like tool search: https://modelcontextprotocol.io/docs/develop/clients/client-best-practices#progressive-tool-discovery

#### Notifications

> The protocol supports real-time notifications to enable dynamic updates between servers and clients. For example, when a server's available tools change—such as when new functionality becomes available or existing tools are modified—the server can send tool update notifications to inform connected clients about these changes. Notifications are sent as JSON-RPC 2.0 notification messages (without expecting a response). Change notifications are opt-in: the client opens a long-lived [`subscriptions/listen`](/specification/draft/basic/patterns/subscriptions) stream naming the notification types it wants to receive, and the server delivers matching notifications on that stream.

Reviewer 2: Two em dashes on this line, parentheses read cleaner.

```suggestion
The protocol supports real-time notifications to enable dynamic updates between servers and clients. For example, when a server's available tools change (such as when new functionality becomes available or existing tools are modified), the server can send tool update notifications to inform connected clients about these changes. Notifications are sent as JSON-RPC 2.0 notification messages (without expecting a response). Change notifications are opt-in: the client opens a long-lived [`subscriptions/listen`](/specification/draft/basic/patterns/subscriptions) stream naming the notification types it wants to receive, and the server delivers matching notifications on that stream.
```

- `"resources": {}` - The server also supports the resources primitive (can handle `resources/list` and `resources/read` methods)

Calling `server/discover` is optional. Because every request carries the same `_meta` fields, a client is free to send any request directly and handle a version error if one comes back. Discovery is a convenient way to fetch the server's identity, capabilities, and supported versions in a single request.

#### How This Works in AI Applications

The AI application's MCP client manager connects to configured servers and stores their discovered capabilities for later use. The application uses this information to determine which servers can provide specific types of functionality (tools, resources, prompts) and whether they support real-time updates.

```python Pseudo-code for AI application discovery
# Pseudo Code
async with stdio_client(server_config) as (read, write):
    async with ClientSession(read, write) as session:
        discover_response = await session.discover()
        if discover_response.capabilities.tools:
            app.register_mcp_server(session, supports_tools=True)

>         app.set_server_ready(session)

Reviewer: Since there are no more sessions, using a hypothetical client SDK API with sessions is giving the wrong impression. I'd just replace `ClientSession` with `Client` and `session` with `client`.

(for some reason github won't let me suggest code edit here)

```

</Step>

<Step title="Tool Discovery (Primitives)">

> The client can discover available tools by sending a `tools/list` request. This request is fundamental to MCP's tool discovery mechanism — it allows clients to understand what tools are available on the server before attempting to use them.

Reviewer 2: Em dash, a colon does the job here.

```suggestion
The client can discover available tools by sending a `tools/list` request. This request is fundamental to MCP's tool discovery mechanism: it allows clients to understand what tools are available on the server before attempting to use them.
```

#### Notifications

The protocol supports real-time notifications to enable dynamic updates between servers and clients. For example, when a server's available tools change (such as when new functionality becomes available or existing tools are modified), the server can send tool update notifications to inform connected clients about these changes. Notifications are sent as JSON-RPC 2.0 notification messages (without expecting a response). Change notifications are opt-in: the client opens a long-lived [`subscriptions/listen`](/specification/draft/basic/patterns/subscriptions) stream naming the notification types it wants to receive, and the server delivers matching notifications on that stream.

## Example

### Data Layer

This section provides a step-by-step walkthrough of an MCP client-server interaction, focusing on the data layer protocol. We'll demonstrate discovery, tool operations, and notifications using JSON-RPC 2.0 messages.

<Steps>
<Step title="Discovery">

> As described in the [statelessness and discovery](#statelessness-and-discovery) section, every MCP request carries the protocol version and client capabilities in its `_meta` field, and clients should also include their identity there. A client that wants to learn what a server supports before issuing other requests sends a `server/discover` request, which every server must implement.

Reviewer 3: I want to make it a little more clear that discovery isn't an "every request" kind of thing, it's usually once per server, and even less often than `tools/list`. 

```suggestion
As described in the [statelessness and discovery](#statelessness-and-discovery) section, every MCP request carries the protocol version and client capabilities in its `_meta` field, and clients should also include their identity there. A client that wants to learn what a server supports before issuing other requests sends a `server/discover` request, which every server must implement. The discovery response is typically cacheable, meaning it can be re-used so the discovery flow does not need to be performed for every request.
```

      "serverInfo": {
        "name": "example-server",
        "version": "1.0.0"

>       },

Reviewer 2: Heads up that #3002 on `main` changed this shape: `DiscoverResult` no longer has a top-level `serverInfo`. Server identity now rides in the result's `_meta` as `io.modelcontextprotocol/serverInfo` ("Servers **SHOULD** include this field on every response", per the [`server/discover` page](https://modelcontextprotocol.io/specification/draft/server/discover)).

```suggestion
      "_meta": {
        "io.modelcontextprotocol/serverInfo": {
          "name": "example-server",
          "version": "1.0.0"
        }
      },
```

      "ttlMs": 3600000,
      "cacheScope": "public"
    }
  }
  ```
</CodeGroup>

#### Understanding the Discovery Exchange

The `_meta` fields and the discovery response together serve several purposes:

1. **Protocol Version Selection**: The `io.modelcontextprotocol/protocolVersion` field declares the version the client is speaking on this request, and `supportedVersions` in the response lists the versions the server accepts. If a server does not support the requested version, it rejects the request with an `UnsupportedProtocolVersionError` listing the versions it does support, and the client retries with a mutually supported version.

2. **Capability Discovery**: The client declares its capabilities in `io.modelcontextprotocol/clientCapabilities` on every request, and the server returns its own `capabilities` object from `server/discover`. This tells each party which [primitives](#primitives) the other can handle (tools, resources, prompts) and whether change [notifications](#notifications) are available, so unsupported operations are never attempted.

> 3. **Identity Exchange**: The `io.modelcontextprotocol/clientInfo` and `serverInfo` objects provide identification and versioning information for debugging and compatibility purposes.

Reviewer 2: Same #3002 change: the server half of the identity exchange is `io.modelcontextprotocol/serverInfo` in the result's `_meta`, mirroring `clientInfo` on the request.

```suggestion
3. **Identity Exchange**: The `io.modelcontextprotocol/clientInfo` field in the request's `_meta` and the `io.modelcontextprotocol/serverInfo` field in the result's `_meta` provide identification and versioning information for debugging and compatibility purposes.
```

   - `location`: "San Francisco" (required parameter)
   - `units`: "imperial" (optional parameter, defaults to "metric" if not specified)

> 3. **`_meta`**: Carries the standard per-request fields (protocol version, client identity, client capabilities) that every MCP request includes.

Reviewer 2: Since #3002, `io.modelcontextprotocol/clientInfo` is a SHOULD rather than a required field: the [Basic overview](https://modelcontextprotocol.io/specification/draft/basic/index) marks it Required: No and says clients **SHOULD** include it "unless specifically configured not to do so". Only `protocolVersion` and `clientCapabilities` are required. The statelessness intro earlier on this page carries the same overstatement.

```suggestion
3. **`_meta`**: Carries the standard per-request fields: the protocol version and client capabilities that every MCP request must include, plus the client's identity, which clients should include unless configured not to.
```

</Step>

<Step title="Real-time Updates (Notifications)">
MCP supports real-time notifications that enable servers to inform clients about changes without being polled for them. This demonstrates the notification system, a key feature that keeps clients synchronized and responsive.

#### Understanding Tool List Change Notifications

Change notifications are opt-in. The client opens a long-lived stream with a [`subscriptions/listen`](/specification/draft/basic/patterns/subscriptions) request that names the notification types it wants to receive:

```json Listen Request
{
  "jsonrpc": "2.0",
  "id": 4,
  "method": "subscriptions/listen",
  "params": {
    "_meta": {
      "io.modelcontextprotocol/protocolVersion": "2026-07-28",
      "io.modelcontextprotocol/clientInfo": {
        "name": "example-client",
        "version": "1.0.0"
      },
      "io.modelcontextprotocol/clientCapabilities": {
        "elicitation": {}
      }
    },
    "notifications": {
      "toolsListChanged": true
    }
  }
}
```

> The first message the server sends on the stream is always `notifications/subscriptions/acknowledged`. Its `notifications` field reflects the subset of the requested filter the server agreed to honor, with unsupported notification types omitted:

Reviewer 2: Per the [Subscriptions page](https://modelcontextprotocol.io/specification/draft/basic/patterns/subscriptions#acknowledgment) the ordering is scoped to the subscription ("the first message carrying the subscription's ID in `_meta`"), and on stdio messages for other subscriptions may interleave ahead of it, so "the first message the server sends on the stream is always" overstates it. #3068 rewrites this same paragraph, so worth matching whatever wording lands there:

```suggestion
The server acknowledges the subscription with `notifications/subscriptions/acknowledged`, which is the first message carrying that subscription's ID in `_meta` (the server sends no other notification for that subscription before it). Its `notifications` field reflects the subset of the requested filter the server agreed to honor, with unsupported notification types omitted:
```

```json Acknowledgment
{
  "jsonrpc": "2.0",
  "method": "notifications/subscriptions/acknowledged",
  "params": {
    "_meta": {
      "io.modelcontextprotocol/subscriptionId": 4
    },
    "notifications": {
      "toolsListChanged": true
    }
  }
}
```

> After the acknowledgment, when the server's available tools change—such as when new functionality becomes available, existing tools are modified, or tools become temporarily unavailable—the server delivers a notification on that stream:

Reviewer 2: Same em dash pair as the notifications paragraph above.

```suggestion
After the acknowledgment, when the server's available tools change (new functionality becomes available, existing tools are modified, or tools become temporarily unavailable), the server delivers a notification on that stream:
```
